Sources/SwiftParser/Lexer/Cursor.swift

```diff
@@ -133,11 +133,61 @@ extension Lexer.Cursor {
       case .afterStringLiteral: return false
       case .afterClosingStringQuote: return false
       case .inStringInterpolationStart: return false
-      case .inStringInterpolation(stringLiteralKind: let stringLiteralKind, parenCount: _, stringInterpolationStart: _): return stringLiteralKind != .multiLine
+      case .inStringInterpolation(stringLiteralKind: let stringLiteralKind, parenCount: _): return stringLiteralKind != .multiLine
       case .afterStringInterpolation: return false
       }
     }
   }
+
+  /// A data structure that holds the state stack entries in the lexer. It is
+  /// optimized for situations where there's at most one state on the stack
+  /// (in addition to the bottom `normal` mode) and to not create any ARC
+  /// traffic.
+  ///
+  /// It does this by representing the bottom `normal` state implicitly and
+  /// storing one additional state inline. If the stack needs to contain more
+  /// entries, the required memory is allocated in a bump allocator that is
+  /// expected to outlive the stack.
+  struct StateStack {
+    private var topState: State? = nil
+    private var stateStack: UnsafeBufferPointer<State>? = nil
+
+    var currentState: State {
+      return topState ?? .normal
+    }
+
+    mutating func perform(stateTransition: Lexer.StateTransition, stateAllocator: BumpPtrAllocator) {
+      switch stateTransition {
+      case .push(newState: let newState):
+        if let topState = topState {
+          if let stateStack = stateStack {
+            let newStateStack = stateAllocator.allocate(State.self, count: stateStack.count + 1)
```

Comment:
It's a little unfortunate we don't reuse previously-allocated buffer. For example
`push`, `push`, `pop`, `push`. operation should reuse the buffer allocated at the second `push`, instead of allocating a whole new buffer.

Reply:
Yes, but how do you know that you have already seen this exact stack? We could try to create a dictionary for it, but given that the bump allocator doesn’t allocate more than 250 byte for any file in MovieSwiftUI, I’m not sure if there’s much to be gained here.

Reply:
I am convinced that we can't reuse the memory region since other  (copied) cursor might still have a reference to it.